### PR TITLE
[flutter_tools] Add support for web in plugin template.

### LINKED ISSUE #59562: Create a template for web in the flutter create -t plugin command
The plugin template for `flutter create` is missing a template for web
https://github.com/flutter/flutter/tree/master/packages/flutter_tools/templates/plugin

In order to add web support with the new `add-platform` command, and to support creating plugin for web, we will need to add a template for web in the directory, as well as updating the pubspec.yaml in the directory to support web.

### PR DESCRIPTION
**Pull request by ditman**
## Description

This PR adds support for the web platform with the plugin template in a backwards-compatible way (using MethodChannels).

This PR addresses comments on #59700 (which maybe should be re-done as a new `federated-plugin` template, if needed.)

## Related Issues

* #59562

## Tests

I added the following tests:

In `flutter_tools/test/commands.shard/permeable/create_test.dart`:

* `plugin project supports web`
* Asserted `lib/flutter_project_web.dart` is not created by default.

```
dit@dit:~/github/flutter/packages/flutter_tools$ pub run test test/commands.shard/permeable/create_test.dart
Precompiling executable... (8.9s)
Precompiled test:test.
03:02 +59: All tests passed!

dit@dit:~/github/flutter/packages/flutter_tools$ pub run test test/commands.shard/permeable/create_test.dart -n web
00:21 +1: All tests passed!
```

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart
Comment (cyanglaz) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:410]: The logic that includes web is kinda weird compares to other platforms.
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/template.dart#L112-L114
As this PR doesn't try to change the logic there, let's add another test to ensure the web related files are not created if web is not requested. WDYT?

File: packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl
Comment (cyanglaz) [inline packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:38]: Can we just reuse the same method channel implementation in the <plugin_name>.dart and this web class just listen to the channel message?

Comment (ditman) [reply packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:38]: I don't think so. If I modify the <plugin_name>.dart this script becomes backwards incompatible, right?

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart
Comment (ditman) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:410]: I added `unexpectedPaths` to the normal plugin creation (where by default `web` is not requested):

```dart
      unexpectedPaths: <String>[
        'lib/flutter_project_web.dart',
      ],
```

Isn't that enough?

Comment (cyanglaz) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:410]: oh I missed that. that's good enough thanks.

File: packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl
Comment (cyanglaz) [reply packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:38]: So we keep the <plugin_name>.dart the same. And the `web.dart` file just listens to the same method channel like other platforms do.

Comment (ditman) [reply packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:38]: Correct, in this case the web implementation listens to the same method channel. The fact that they both live right next to each other in the directory hierarchy is confusing. Imagine this `web.dart` lives in a completely separate package.

Comment (cyanglaz) [reply packages/flutter_tools/templates/plugin/lib/projectName_web.dart.tmpl:38]: Again, my mistake. I was looking at something else. This is good. thanks

**Review by cyanglaz — APPROVED** [review]
LGTM! cc @jonahwilliams Could you also take a look?

**Comment by ditman** [comment]
**NOTE:** I've been running this in my machine:

`dit@dit:~/github$ dart flutter/packages/flutter_tools/bin/flutter_tools.dart create -t plugin some_test`

**Without passing `web` anywhere.** Somewhere in my tooling "web" is enabled by default. I'm not sure if this is expected behavior or a bug, but it seems once this gets merged people will start creating "web" plugins by default?

**Comment by jonahwilliams** [comment]
This is expected right now because web plugins are only guarded by the experiment. @cyanglaz is adding platform flags to fix this

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by ditman** [comment]
Thanks @jonahwilliams!

I'll let @cyanglaz merge this PR when he's happy with it, rather than setting the `waiting for CI to go green` label and letting the bot do it before everything is ready.

@cyanglaz, feel free to Squash and merge this whenever you want!

**Comment by ditman** [comment]
Discussed with @cyanglaz on chat, and it's OK that this lands before his changes. I'm setting the Waiting for CI to go green label so it gets merged whenever the tree is ready.

**Comment by flutter-github-sync** [comment]
Google testing passed!

**Pull request merged**
